app/jobs/inherited_proofing_job.rb

```diff
@@ -0,0 +1,20 @@
+class InheritedProofingJob < ApplicationJob
+  include JobHelpers::StaleJobHelper
+
+  queue_as :default
+
+  discard_on JobHelpers::StaleJobHelper::StaleJobError
+
+  def perform(service_provider, service_provider_data, uuid)
+    document_capture_session = DocumentCaptureSession.find_by(uuid: uuid)
+
+    payload_hash = Idv::InheritedProofing::ServiceProviderServiceFactory.execute(
```

Comment:
I'm not sure I understand the benefit we get from employing factory objects here? I don't see where it simplifies things but it does seem to add another layer of indirection and makes it difficult for me to understand exactly what this code is doing or what `payload_hash` is.

Reply:
Hi @jmhooper, I agree it does make things a little more complicated unfortunately. The code to identify the Service Provider (SP)-specific service API needed to obtain user PII, as well as the code needed to identify the SP-specific Form necessary to consume the SP-specific service result (`payload_hash`), was originally in the `InheritedProofingConcern` which was bound to Session, along with all the code to access SP-specific API services and SP-specific forms. This is because (currently) the only way to identify the Service Provider (SP) we're dealing with, is by querying the Session (`decorated_session`). Unless ActiveJob has a way to access Session, we have no way to identify the SP except by passing it. Regardless, the code had to be refactored so that it could be mixed into `ActiveJob` (call the SP-specific service), the `InheritedProofingController` for the IP Flow Steps (identify SP-specific Form to consume `payload_hash`), and (eventually) the code needed to clean up the IP session if the user decides to bag out of the IP process. All this had to be done in an SP-agnostic way, and be "carved up" to be used in each of these scenarios.

The code is really not that complicated (just a little _less_ simple), but I see no other way without specifically referencing SP's in a lot of places, and having an equal amount of control statements once other SP's come aboard, and a _lot_ of refactoring.

Reply:
I didn't have time to comment that code unfortunately; but, I will go back and add comments.

Reply:
I am a strong agree with Hooper here, I think that given that there is only one of these things, a Factory is not useful, and a Factory we subclass is even harder to follow. I vote to collapse all that code if we can.

Reply:
Dealt with in this PR: https://github.com/18F/identity-idp/pull/7181